### DISCUSSION TITLE
Some DX recommendations for monorepos

### DISCUSSION THREAD

**Discussion by zomars** [trpc/trpc#2132] [Q&A] [answered]
Coming from [this list](https://github.com/t3-oss/create-t3-app/issues/140#issuecomment-1178876716)

At [calcom/cal.com](https://github.com/calcom/cal.com) we're looking into migrating tRPC into it's own package so we can use it to server both the public API and web app among some other packages.

What would be some recommendations for setup/testing and dev experience overall?

**Comment by ixahmedxi** [comment]
Idk how helpful this is but [here](https://github.com/ixahmedxi/noodle/tree/main/libs/server) I have a separate package for trpc and the API, I export the main router and the router type and then that gets used to make let's say a nextjs api handler on the next.js app. This gives me the flexibility to plug in my trpc API to any kind of server from nextjs to express if I needed and decouples the trpc API from any framework.

**Comment by KATT** [comment, marked as answer]
Hey @zomars, thanks for your support of tRPC :)

I know that some people use internally published npm packages to share types between services, but I personally prefer the approach that cal is taking with having everything in the same repo and using something like Turborepo or Nx to orchestrate it, internally published packages quickly become annoying with versioning and whatnot.

I'm not an expert on any of the modern monorepo tools _(tRPC itself still uses Lerna 👴)_ but I think a good thing to do is to follow the best practises outlined for each. 

Some things that can help you though:

- As you already noted, read through my general DX-comments on the "t3-stack" [here](https://github.com/t3-oss/create-t3-app/issues/140#issuecomment-1178876716).
- You might want to do some more advanced env parsing since you have several integrations in your setup that can be opted - [see my comment here](https://github.com/calcom/cal.com/issues/3261#issuecomment-1179522116)
- Utilize the caching and build optimizations from the tools in order to only build & test changed files and their dependents - [Turborepo docs](https://turborepo.org/docs/core-concepts/caching) 
- Use Playwright for E2E testing to make sure it all works together. I know Playwright quickly becomes a bottleneck in CI and I'm unsure how to properly optimize the time it takes.
- Try to reuse the same tsconfig/eslint/etc settings across each package to ensure consistency, tsconfig has a way of [`extend`ing](https://www.typescriptlang.org/tsconfig/extends.html) other configs. Potentially tsconfigs and things could be automatically generated (and gitignored) from a `postinstall`-script in the root of the config if there are many many packages within the app to avoid inconsistencies.
- In general, I think [the cal monorepo structure](https://github.com/calcom/cal.com) looks really good  
  - The fact that the full project is _partially_ open-source adds a layer of complexity to the whole setup and has great DX which I don't know how to deal with properly. 
    - If you haven't already, I'd reach out to similar companies to yourselves and see how they deal with this effectively and probably have someone spend a few days with the whole setup and how you sync between private<->public.
    - I'm unsure if it's best to _sync_ (having copies) or to leverage submodules as you do now. It might be best to have a private "full monorepo" + a public "partial monorepo" and automatically sync between them when there's a push to either (subtree might help here, see below). I know [Jonatan](https://twitter.com/jonatansberg) has done some really cool stuff with subtree, maybe hire his company for assistance could be an idea?
  - I don't really have a good personal experience with git submodules which you seem to be relying on:
    - Can a person who doesn't have access to the private git repos effectively contribute to the project? I'd probably add some automated testing somehow to assert that this notion remains true
    - I've played a bit with [git subtree](https://www.atlassian.com/git/tutorials/git-subtree) (that's how I keep [the examples](https://github.com/trpc/trpc/tree/main/examples/next-prisma-starter) in sync with [their designated repos](https://github.com/trpc/examples-next-prisma-starter) - [CI job here](https://github.com/trpc/trpc/blob/main/.github/workflows/subtree.yml))
  - Maybe the "api" should be a package rather than an app in order to be able to reuse the types across several apps?

**Reply by JacobMGEvans** [reply]
> Use Playwright for E2E testing to make sure it all works together. I know Playwright quickly becomes a bottleneck in CI and I'm unsure how to properly optimize the time it takes.

Caching stuff is almost always the answer to CI perf lmao

**Reply by JacobMGEvans** [reply]
> tRPC itself still uses Lerna 

but but... TurboRepo is soooooo gooood 😆  

The rewrite is likely not worth it right now in all seriousness but could be eventually.
